api/src/main/java/org/jenkinsci/plugins/workflow/graph/FlowNode.java

```diff
@@ -103,6 +115,16 @@ public final boolean isRunning() {
      * Returns a read-only view of parents.
      */
     public List<FlowNode> getParents() {
+        if (parents == null) {
+            parents = new ArrayList<FlowNode>(parentIds.size());
+            for (String id : parentIds) {
```

Comment:
Should only be used during deserialization, not randomly during runtime. Synchronizing this seems more dangerous than not. Unsure if there is some practical way to make this safer.